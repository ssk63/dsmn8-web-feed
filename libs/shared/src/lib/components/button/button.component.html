<button
  class="dsmn8-btn"
  [class.dsmn8-btn--primary]="isPrimary"
  [attr.aria-label]="ariaLabel"
  type="button"
  *ngIf="label"
>
  {{ label }}
</button>
