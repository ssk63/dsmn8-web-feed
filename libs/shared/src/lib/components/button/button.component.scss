$primary: #e85427;
$primary-light: #fb7147;
$grey: #455a76;
$white: #ffffff;

.dsmn8-btn {
  &--primary {
    border-color: $primary;
    background-color: $primary;
    color: $white;
    &:hover {
      border-color: $primary-light;
      background-color: $primary-light;
    }
  }
  &--block {
    width: 100%;
  }
  &[disabled] {
    background-color: $grey;
    border-color: $grey;
    color: $white;
    opacity: 0.38;
    cursor: default;
  }
  border: solid 1px #eceef1;
  border-radius: 10px;
  background-color: $white;
  color: #494949;
  padding: 12px 30px;
  margin-bottom: 20px;
  transition-duration: 0.5s;
  text-align: center;
  cursor: pointer;
  min-width: 160px;
}
