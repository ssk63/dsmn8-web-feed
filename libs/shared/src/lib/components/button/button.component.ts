import {
  Component,
  EventEmitter,
  Input,
  Output,
  ViewEncapsulation,
} from '@angular/core';

@Component({
  selector: 'dsmn8-button',
  templateUrl: './button.component.html',
  styleUrls: ['./button.component.scss'],
  encapsulation: ViewEncapsulation.None,
})
export class ButtonComponent {
  @Input() label: string | undefined;
  @Input() ariaLabel: string | undefined;
  @Input() isPrimary: boolean | undefined;
  @Input() isDisabled: boolean | undefined;
  @Output() clicked = new EventEmitter<boolean>();

  onBtnClick(): void {
    this.clicked.emit(true);
  }
}
