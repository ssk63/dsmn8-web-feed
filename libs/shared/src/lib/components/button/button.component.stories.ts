import { moduleMetadata, Story, Meta } from '@storybook/angular';
import { withDesign } from 'storybook-addon-designs';
import { ButtonComponent } from './button.component';

export default {
  title: 'Shared/Components/ButtonComponent',
  component: ButtonComponent,
  decorators: [
    withDesign,
    moduleMetadata({
      imports: [],
    }),
  ],
  argTypes: {
    label: {
      name: 'label',
      description: 'Sets the button text.',
      type: { name: 'string', required: true },
      control: {
        type: 'text',
      },
      table: {
        category: 'Inputs',
      },
    },
    isDisabled: {
      name: 'isDisabled',
      description: 'Whether the button is disabled.',
      type: {
        name: 'boolean',
      },
      table: {
        category: 'Inputs',
      },
    },
    isPrimary: {
      name: 'isPrimary',
      description: 'Whether the button is primary.',
      type: {
        name: 'boolean',
      },
      table: {
        category: 'Inputs',
      },
    },
    isFullWidth: {
      name: 'isFullWidth',
      description:
        'If true, the button stretches to the full width of its container.',
      defaultValue: false,
      type: {
        name: 'boolean',
      },
      table: {
        category: 'Inputs',
      },
    },
    clicked: {
      description: 'Event that is emitted when an button is clicked.',
      action: 'clicked',
      table: {
        category: 'Outputs',
      },
    },
  },
  parameters: {
    jest: 'button.component',
  },
} as Meta<ButtonComponent>;

const Template: Story<ButtonComponent> = (args: ButtonComponent) => ({
  props: {
    ...args,
    // clicked: action('log'),
  },
});

export const Primary = Template.bind({});
Primary.args = {
  label: 'Button',
  isPrimary: true,
  isDisabled: false,
};
Primary.parameters = {
  design: {
    type: 'figma',
    url: 'https://www.figma.com/file/Klm6pxIZSaJFiOMX5FpTul9F/storybook-addon-designs-sample',
    allowFullscreen: true,
  },
};

export const Disabled = Template.bind({});
Disabled.args = {
  ...Primary.args,
  isDisabled: true,
  isPrimary: false,
};
Disabled.parameters = {
  design: {
    type: 'figma',
    url: 'https://www.figma.com/proto/Klm6pxIZSaJFiOMX5FpTul9F/storybook-addon-designs-sample',
  },
};

export const FullWidth = Template.bind({});
FullWidth.args = {
  ...Primary.args,
  isFullWidth: true,
};
FullWidth.parameters = {
  design: [
    {
      name: 'Figma',
      type: 'figma',
      url: 'https://www.figma.com/file/Klm6pxIZSaJFiOMX5FpTul9F/storybook-addon-designs-sample',
      allowFullscreen: true,
    },
    {
      name: 'Link',
      type: 'link',
      url: 'https://www.figma.com/file/Klm6pxIZSaJFiOMX5FpTul9F/storybook-addon-designs-sample',
    },
  ],
};
