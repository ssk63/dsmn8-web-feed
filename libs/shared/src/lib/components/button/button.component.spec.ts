import { ComponentFixture, TestBed } from '@angular/core/testing';
import { By } from '@angular/platform-browser';
import { ButtonComponent } from './button.component';

describe('ButtonComponent', () => {
  let component: ButtonComponent;
  let fixture: ComponentFixture<ButtonComponent>;

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      declarations: [ButtonComponent],
    }).compileComponents();

    fixture = TestBed.createComponent(ButtonComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  test('should render button when the label is exists', () => {
    // arrange
    component.label = 'Like Button';
    // act
    fixture.detectChanges();
    const button = fixture.debugElement.query(
      By.css('.dsmn8-btn')
    ).nativeElement;
    // assert
    expect(button).toMatchSnapshot();
  });

  test('should not render button when the label is empty', () => {
    // arrange
    component.label = '';
    // act
    fixture.detectChanges();
    const button = fixture.debugElement.query(By.css('.dsmn8-btn'));
    // assert
    expect(button).toBeNull();
  });

  test('should be called onBtnClick() method when the button is clicked', () => {
    // arrange
    component.label = 'Share button';
    jest.spyOn(component, 'onBtnClick');
    // act
    fixture.detectChanges();
    const button = fixture.debugElement.query(By.css('.dsmn8-btn'));
    button.nativeElement.click();
    // assert
    expect(component.onBtnClick).toBeCalled();
  });

  test('should emit clicked event true when the onBtnClick() method is called', () => {
    // arrange
    jest.spyOn(component.clicked, 'emit');
    // act
    component.onBtnClick();
    // assert
    expect(component.clicked.emit).toBeCalledWith(true);
  });

  test('should render primary button when the isPrimary is true', () => {
    // arrange
    component.isPrimary = true;
    component.label = 'Like Button';
    // act
    fixture.detectChanges();
    const primaryButton = fixture.debugElement.query(
      By.css('.dsmn8-btn--primary')
    ).nativeElement;
    // assert
    expect(primaryButton).toMatchSnapshot();
  });

  test('should render fullWidth button when the isFullWidth is true', () => {
    // arrange
    component.isFullWidth = true;
    component.isPrimary = true;
    component.label = 'Like Button';
    // act
    fixture.detectChanges();
    const fullWidthButton = fixture.debugElement.query(
      By.css('.dsmn8-btn--block')
    ).nativeElement;
    // assert
    expect(fullWidthButton).toMatchSnapshot();
  });

  test('should render disabled button when the isDisabled is true', () => {
    // arrange
    component.isDisabled = true;
    component.isPrimary = true;
    component.label = 'Like Button';
    // act
    fixture.detectChanges();
    const button = fixture.debugElement.query(
      By.css('.dsmn8-btn')
    ).nativeElement;
    // assert
    expect(button.disabled).toBeTruthy();
    expect(button).toMatchSnapshot();
  });
});
