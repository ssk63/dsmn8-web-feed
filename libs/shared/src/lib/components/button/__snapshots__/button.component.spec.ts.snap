// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonComponent should render button when the label is exists 1`] = `
<button
  class="dsmn8-btn"
  type="button"
>
   Like Button

</button>
`;

exports[`ButtonComponent should render disabled button when the isDisabled is true 1`] = `
<button
  class="dsmn8-btn dsmn8-btn--primary"
  disabled=""
  type="button"
>
   Like Button

</button>
`;

exports[`ButtonComponent should render fullWidth button when the isFullWidth is true 1`] = `
<button
  class="dsmn8-btn dsmn8-btn--primary dsmn8-btn--block"
  type="button"
>
   Like Button

</button>
`;

exports[`ButtonComponent should render primary button when the isPrimary is true 1`] = `
<button
  class="dsmn8-btn dsmn8-btn--primary"
  type="button"
>
   Like Button

</button>
`;
